hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/BlockStreamManagerImpl.java

```diff
@@ -0,0 +1,400 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.records.impl;
+
+import static com.hedera.node.app.records.BlockRecordService.EPOCH;
+import static com.hedera.node.app.records.impl.BlockRecordInfoUtils.HASH_SIZE;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.Timestamp;
+import com.hedera.hapi.node.state.blockrecords.BlockInfo;
+import com.hedera.hapi.node.state.blockrecords.RunningHashes;
+import com.hedera.node.app.records.BlockRecordManager;
+import com.hedera.node.app.records.BlockRecordService;
+import com.hedera.node.app.records.schemas.V0490BlockRecordSchema;
+import com.hedera.node.app.state.SingleTransactionStreamRecord;
+import com.hedera.node.config.ConfigProvider;
+import com.hedera.node.config.data.BlockStreamConfig;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.common.crypto.DigestType;
+import com.swirlds.common.crypto.Hash;
+import com.swirlds.platform.state.PlatformState;
+import com.swirlds.platform.state.spi.WritableSingletonStateBase;
+import com.swirlds.platform.system.transaction.ConsensusTransaction;
+import com.swirlds.state.HederaState;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Stream;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ *
+ */
+@Singleton
+public class BlockStreamManagerImpl implements BlockRecordManager {
+    private static final Logger logger = LogManager.getLogger(BlockStreamManagerImpl.class);
+    private final ExecutorService executor;
+
+    /**
+     * The number of rounds in the current provisional block. "provisional" because the block is not yet complete.
+     */
+    // Should this be stored in BlockInfo so that it's safe for restarts?
+    private int roundsUntilNextBlock;
+
+    /**
+     * Keeps track if this block is currently open.
+     */
+    // Should this be stored in BlockInfo so that it's safe for restarts?
+    private boolean blockOpen;
+
+    private final int numBlockHashesToKeepBytes;
+    private final int numRoundsInBlock;
+    private final BlockStreamProducer blockStreamProducer;
+    /** True when we have completed event recovery. This is not yet implemented properly. */
+    private boolean eventRecoveryCompleted = false;
+
+    /**
+     * A {@link BlockInfo} of the most recently completed block. This is actually available in state, but there
+     * is no reason for us to read it from state every time we need it, we can just recompute and cache this every
+     * time we finish a provisional block.
+     */
+    private BlockInfo lastBlockInfo;
+
+    @Inject
+    public BlockStreamManagerImpl(
```

Comment:
Does this class need to be thread safe? I don't think so since we should only be calling it from the handle workflow, but it's worth asking the question